internal/query/map.go

```diff
@@ -1,81 +1,198 @@
 package query
 
 import (
+	"errors"
 	"fmt"
 	"strings"
 )
 
+const MaxNestedMapDepth = 100
+
+type queryKeyType int
+
+const (
+	filteredUnsupported queryKeyType = iota
+	filteredMap
+	filteredRejected
+	mapType
+	arrayType
+	emptyKeyValue
+	valueType
+)
+
 // GetMap returns a map, which satisfies conditions.
-func GetMap(query map[string][]string, key string) (dicts map[string]interface{}, exists bool) {
+func GetMap(query map[string][]string, filteredKey string) (map[string]interface{}, error) {
 	result := make(map[string]interface{})
-	for qk, value := range query {
-		if isKey(qk, key) {
-			path, err := parsePath(qk, key)
+	getAll := filteredKey == ""
+	var allErrors = make([]error, 0)
+	for key, value := range query {
+		kType := getType(key, filteredKey, getAll)
+		switch kType {
+		case filteredUnsupported:
+			allErrors = append(allErrors, fmt.Errorf("invalid access to map %s", key))
+			continue
+		case filteredMap:
+			fallthrough
+		case mapType:
+			path, err := parsePath(key)
+			if err != nil {
+				allErrors = append(allErrors, err)
+				continue
+			}
+			if !getAll {
+				path = path[1:]
+			}
+			err = setValueOnPath(result, path, value)
+			if err != nil {
+				allErrors = append(allErrors, err)
+				continue
+			}
+		case arrayType:
+			err := setValueOnPath(result, []string{keyWithoutArraySymbol(key), ""}, value)
+			if err != nil {
+				allErrors = append(allErrors, err)
+				continue
+			}
+		case filteredRejected:
+			continue
+		case emptyKeyValue:
+			result[key] = value[0]
+		default:
+			err := setValueOnPath(result, []string{key}, value)
 			if err != nil {
+				allErrors = append(allErrors, err)
 				continue
 			}
-			setValueOnPath(result, path, value)
-			exists = true
 		}
 	}
-	if !exists {
-		return nil, exists
+	if len(allErrors) > 0 {
+		return nil, errors.Join(allErrors...)
+	} else if len(result) == 0 {
+		return nil, nil
 	}
-	return result, exists
+	return result, nil
+}
 
+// getType is an internal function to get the type of query key.
+func getType(key string, filteredKey string, getAll bool) queryKeyType {
+	if getAll {
+		if isMap(key) {
+			return mapType
+		} else if isArray(key) {
+			return arrayType
+		} else if key == "" {
+			return emptyKeyValue
+		}
+		return valueType
+	}
+	if isFilteredKey(key, filteredKey) {
+		if isMap(key) {
+			return filteredMap
+		}
+		return filteredUnsupported
+	}
+	return filteredRejected
+}
```

Comment:
Maybe skip the `else` part because `returns` already break the function execution. 
- Of if you really like `elses` in such cases - be consistent and use this approach for all the `if-else` places.